### PR TITLE
Remove Flux from lib/infinite-list scroll positions store

### LINKED ISSUE #24178: Reduxify: infinite-list
**summary**
`infinite-list` depends on two flux store, [scroll-store](https://github.com/Automattic/wp-calypso/blob/master/client/lib/infinite-list/scroll-store.js) and [positions-store](https://github.com/Automattic/wp-calypso/blob/master/client/lib/infinite-list/positions-store.js).
Both of those stores are _only_ used within the infinite-list component.
Both stores are glorified setters and getters. very little complexity. Only two actions, one per store: `storeScroll`, `storePosition`

| task size | files      | LOC |
| ----------- | ---------- | ----|
| small      | 3 | 130 |

### PR DESCRIPTION
**Pull request by jsnajdr**
Replaces the Flux store in `lib/infinite-list` with a very simple implementation using `Map` and getter and (throttled) setter functions.

The store only has a getter/setter API, and no change events that anyone would subscribe to. That's why it doesn't even need Redux.

**How to test:**
- enable debug logging by enabling it in console and reloading:
```js
localStorage.debug = 'calypso:infinite-list:scroll-store'
```
- go to a route that uses `InfiniteList`, e.g., the media library
- scroll down in the list of media
- go to another page
- click the "Back" button and navigate back to media library
- check that when navigating using browser history, the scroll position is reconstructed to the previous value
- while testing, watch the debug messages in console and verify there's nothing unexpected going on

@blowery added some Reader-related functionality to `InfiniteList` in the past. He could know some more sophisticated testing steps, especially for the Reader Stream usage 🙂 

Fixes #24178.

### CODE REVIEW

**Comment by samouri** [comment]
> The store only has a getter/setter API, and no change events that anyone would subscribe to. That's why it doesn't even need Redux.

I had the exact same thought when writing #24178 :)
Reviewing this now...

File: client/components/infinite-list/scroll-store.js
Comment (samouri) [inline client/components/infinite-list/scroll-store.js]: double checking: this means it takes the latest call right instead of the first call right?

File: client/components/infinite-list/scroll-store.js
Comment (samouri) [inline client/components/infinite-list/scroll-store.js]: perfect example of why flux/redux can be over-engineering!

**Comment by samouri** [comment]
> (throttled) setter functions.

I was curious about this.  Change Events can be expensive, which is a good reason to throttle something like this.  With the new implementation, it seems to me like the `throttle` is either as expensive or _more expensive_ than calling `map.set`.  Is there a reason to maintain the throttle?

**Comment by samouri** [comment]
I couldn't find any issues during testing.  What I did was scroll like a madman in Reader and then click back and make sure I was brought to the right place. Worked every time.  I even clicked into a post, scrolled with keyboard a few times, and then hit back until returning to the stream. It all still lgtm.

I know @blowery has a borderline clinical need to get scroll positions perfect, so lets wait on his approval before 🚢

File: client/components/infinite-list/scroll-store.js
Comment (jsnajdr) [reply client/components/infinite-list/scroll-store.js]: As I understand `throttle`, the default behavior with `leading: true` is to call the throttled function on the very first scroll event and then call it no earlier than after 1000ms. With `leading: false`, the first call happens 1000ms after the first scroll event.

But it doesn't really matter, as you made a good point that the throttling is not needed at all if there are no change events and listeners.

Comment (jsnajdr) [reply client/components/infinite-list/scroll-store.js]: Hmm, now when I'm testing the un-throttled `store` methods, it breaks the scroll position storing functionality. The `scrollTop` stored value is reset to 0 when navigating away from a route like `/media`. The `InfiniteList` scrolling code seems to be not 100% correct. But I don't want to go down that rabbit hole now: let's just de-Fluxify the stores.

File: client/components/infinite-list/index.jsx
Comment (blowery) [inline client/components/infinite-list/index.jsx]: yay

**Review by blowery — APPROVED** [review]
LGTM! Tested out fine.

I'm not sure why we throttle storage. But, future PR.

**Comment by jsnajdr** [comment]
> I'm not sure why we throttle storage. But, future PR.

There is one buggy behavior I noticed: the URL that is used as a key for storing the scrolling info is taken from `page.current`, where `page` is the router library. During navigation it can happen that `page.current` is already the new route, but the old UI is still mounted. Then the Media Library, for example, will store scrolling data under the wrong key, probably when unmounting or when processing some Redux state update.

This bug and the throttled store calls interact in such a way that everything works correctly 😃 But if you remove the throttling, scroll offset with zero value will be stored when navigating away and won't be correctly restored when navigating back.

This would require very careful cleanup of the `InfiniteList` component code, I'm just writing my findings down here so that they are not lost.

**Comment by blowery** [comment]
> This bug and the throttled store calls interact in such a way that everything works correctly 😃

[Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony) has been released!

**Pull request merged**
